File: apps/android/app/src/main/java/app/inkuna/android/ui/reader/engine/EnginePagerSurface.kt
package app.inkuna.android.ui.reader.engine

import app.inkuna.android.ui.reader.ReaderPagerStrip
import app.inkuna.android.ui.reader.ReaderPagerSurface
import app.inkuna.core.ChapterGeometry
import app.inkuna.core.InkunaException
import app.inkuna.core.ReaderSession
import kotlin.math.roundToInt

sealed interface ChapterReadiness {
    data object Empty : ChapterReadiness
    data class Partial(val publishedPages: UInt) : ChapterReadiness
    data class Complete(val geometry: ChapterGeometry) : ChapterReadiness
}

/** Engine-backed pager state. Every session read is synchronous and cache-only. */
class EnginePagerSurface(
    private val session: ReaderSession,
    private val canvas: EnginePageCanvas,
) : ReaderPagerSurface {
    private data class NeighborKey(val spineIdx: UInt, val toRight: Boolean)

    private val readiness = mutableMapOf<UInt, ChapterReadiness>()
    private val failedSpines = mutableSetOf<UInt>()
    private val neighborReadiness = mutableMapOf<NeighborKey, Boolean>()
    private var latestGeneration: ULong? = null
    private var pendingGeneration = false
    private var innerOffset = 0f
    private var outerDisplacement = 0f
    private var scenePageCount = 0u
    private var interactionPageCount: UInt? = null
    private var lastSettledPage: UInt? = null

    var spineIdx: UInt = 0u
        private set
    var pageIdx: UInt = 0u
        private set
    var spineCount: UInt = session.spineCount()
    var onPageSettled: ((UInt, UInt) -> Unit)? = null
    var selectionActive: Boolean = false

    init {
        canvas.attach(session)
    }

    override val isEngageable: Boolean
        get() = !isBusy && canvas.width > 0 && canvas.height > 0 &&
            (spineIdx in failedSpines || canvas.hasPage(spineIdx, pageIdx))
    override val isBusy: Boolean get() = pendingGeneration
    override val hasActiveSelection: Boolean get() = selectionActive
    override val isRightToLeft: Boolean get() = session.isRtl()

    fun display(spineIdx: UInt, pageIdx: UInt) {
        this.spineIdx = spineIdx
        this.pageIdx = pageIdx
        interactionPageCount = null
        primeReadiness(spineIdx)
        val count = pageCountFor(spineIdx)
        innerOffset = PageSlot.offset(pageIdx, count, isRightToLeft, pageWidth)
        outerDisplacement = 0f
        scenePageCount = count
        lastSettledPage = pageIdx
        neighborReadiness.clear()
        canvas.showUnreadablePlaceholder(spineIdx in failedSpines)
        setScene()
        onPageSettled?.invoke(spineIdx, pageIdx)
    }

    fun firstPageBecameReady(generation: ULong, spineIdx: UInt) {
        if (!accept(generation)) return
        val published = maxOf(1u, session.publishedPageCount(spineIdx))
        if ((readiness[spineIdx] as? ChapterReadiness.Complete)?.geometry?.generation != generation) {
            readiness[spineIdx] = ChapterReadiness.Partial(published)
        }
        failedSpines.remove(spineIdx)
        neighborReadiness.clear()
        if (spineIdx == this.spineIdx) {
            pendingGeneration = false
            canvas.showUnreadablePlaceholder(false)
            canvas.invalidate(generation)
            setScene()
        }
    }

    fun chapterBecameReady(generation: ULong, spineIdx: UInt) {
        if (!accept(generation)) return
        val geometry = try {
            session.chapter(spineIdx)
        } catch (_: InkunaException.NotReady) {
            return
        } catch (_: InkunaException.UnsupportedContent) {
            return
        }
        if (geometry.generation != generation) return
        readiness[spineIdx] = ChapterReadiness.Complete(geometry)
        failedSpines.remove(spineIdx)
        neighborReadiness.clear()
        if (spineIdx == this.spineIdx) {
            pendingGeneration = false
            canvas.showUnreadablePlaceholder(false)
            canvas.invalidate(generation)
            setScene()
        }
    }

    fun chapterFailed(generation: ULong, spineIdx: UInt) {
        if (!accept(generation)) return
        readiness[spineIdx] = ChapterReadiness.Empty
        failedSpines += spineIdx
        neighborReadiness.clear()
        if (spineIdx == this.spineIdx) {
            pendingGeneration = false
            canvas.showUnreadablePlaceholder(true)
            setScene()
        }
    }

    fun layoutInvalidated(generation: ULong) {
        latestGeneration = generation.takeIf { it != 0uL }
        pendingGeneration = true
        readiness.clear()
        failedSpines.clear()
        neighborReadiness.clear()
        scenePageCount = 0u
        interactionPageCount = null
        lastSettledPage = null
        latestGeneration?.let(canvas::invalidate) ?: canvas.invalidateAll()
    }

    override fun beginPagingInteraction() {
        val count = pageCountFor(spineIdx)
        rebaseInnerOffset(count)
        interactionPageCount = count
    }

    override fun endPagingInteraction() {
        interactionPageCount = null
        setScene()
    }

    override fun innerMetrics(): ReaderPagerStrip? {
        val width = pageWidth
        if (width <= 0f) return null
        val count = displayedPageCount()
        rebaseInnerOffset(count)
        val end = (maxOf(1u, count) - 1u).toFloat() * width
        return when (val state = readiness[spineIdx] ?: ChapterReadiness.Empty) {
            is ChapterReadiness.Complete -> {
                if (state.geometry.generation != latestGeneration) return null
                ReaderPagerStrip(innerOffset, 0f..end, width)
            }
            is ChapterReadiness.Partial ->
                ReaderPagerStrip(innerOffset, 0f..maxOf(innerOffset, end), width)
            ChapterReadiness.Empty -> if (spineIdx in failedSpines) {
                ReaderPagerStrip(innerOffset, 0f..0f, width)
            } else {
                null
            }
        }
    }

    override fun setInnerOffset(x: Float) {
        val metrics = innerMetrics() ?: return
        innerOffset = x.coerceIn(metrics.range.start, metrics.range.endInclusive)
        setScene()
        val count = displayedPageCount()
        val slot = (innerOffset / metrics.pageWidth).roundToInt().toUInt()
        val settled = PageSlot.pageIdx(slot, count, isRightToLeft)
        val settledOffset = PageSlot.offset(settled, count, isRightToLeft, metrics.pageWidth)
        if (kotlin.math.abs(innerOffset - settledOffset) < 0.01f && settled != lastSettledPage) {
            pageIdx = settled
            lastSettledPage = settled
            onPageSettled?.invoke(spineIdx, settled)
        }
    }

    override fun outerMetrics(): ReaderPagerStrip? {
        val width = pageWidth
        if (width <= 0f) return null
        val leftExists = neighborSpine(toRight = false) != null
        val rightExists = neighborSpine(toRight = true) != null
        return ReaderPagerStrip(
            offset = width - outerDisplacement,
            range = (if (leftExists) 0f else width)..(if (rightExists) width * 2f else width),
            pageWidth = width,
        )
    }

    override fun setOuterOffset(x: Float) {
        val metrics = outerMetrics() ?: return
        val offset = x.coerceIn(metrics.range.start, metrics.range.endInclusive)
        outerDisplacement = metrics.pageWidth - offset
        setScene()
    }

    override fun neighborIsReady(toRight: Boolean): Boolean {
        val neighbor = neighborSpine(toRight) ?: return false
        val key = NeighborKey(neighbor, toRight)
        neighborReadiness[key]?.let { return it }
        val ready = when {
            // A failed chapter occupies one placeholder page; it must stay
            // crossable in both directions or every chapter beyond it
            // becomes unreachable by paging.
            neighbor in failedSpines -> true
            isForward(toRight) -> {
                val published = when (readiness[neighbor]) {
                    is ChapterReadiness.Partial, is ChapterReadiness.Complete -> true
                    ChapterReadiness.Empty, null -> session.publishedPageCount(neighbor) > 0u
                }
                // A cache miss schedules the chapter, so a later layout
                // event resolves the crossing (mirrors iOS).
                if (!published) runCatching { session.chapter(neighbor) }
                published
            }
            else -> completeGeometry(neighbor) != null
        }
        neighborReadiness[key] = ready
        return ready
    }

    override fun commitBoundaryCrossing(toRight: Boolean): Boolean {
        val target = neighborSpine(toRight) ?: return false
        if (!neighborIsReady(toRight)) return false
        val targetPage = when {
            target in failedSpines -> 0u
            isForward(toRight) -> 0u
            else -> {
                val geometry = completeGeometry(target) ?: return false
                geometry.pageCount - 1u
            }
        }
        spineIdx = target
        pageIdx = targetPage
        val count = pageCountFor(target)
        innerOffset = PageSlot.offset(targetPage, count, isRightToLeft, pageWidth)
        scenePageCount = count
        interactionPageCount = count
        outerDisplacement = 0f
        lastSettledPage = targetPage
        neighborReadiness.clear()
        canvas.showUnreadablePlaceholder(target in failedSpines)
        setScene()
        onPageSettled?.invoke(target, targetPage)
        return true
    }

    private val pageWidth: Float get() = canvas.width.toFloat()

    private fun accept(generation: ULong): Boolean {
        val known = latestGeneration
        if (known != null && known != generation) return false
        if (known == null) {
            latestGeneration = generation
            canvas.invalidate(generation)
        }
        return true
    }

    /**
     * Learns a spine's readiness from the session cache when no layout
     * event has taught it yet — a display can land before the event
     * stream reaches this surface (fresh mount, replayed-out events).
     * Without a page count the strip would anchor every page at slot 0.
     */
    private fun primeReadiness(spineIdx: UInt) {
        if (readiness[spineIdx] != null || spineIdx in failedSpines) return
        val published = session.publishedPageCount(spineIdx)
        if (published > 0u) readiness[spineIdx] = ChapterReadiness.Partial(published)
    }

    private fun pageCountFor(spineIdx: UInt): UInt {
        if (spineIdx in failedSpines) return 1u
        return when (val state = readiness[spineIdx]) {
            is ChapterReadiness.Complete -> state.geometry.pageCount
            is ChapterReadiness.Partial -> maxOf(state.publishedPages, session.publishedPageCount(spineIdx))
            ChapterReadiness.Empty, null -> 0u
        }
    }

    /**
     * Complete geometry with at least one page, healing the readiness map
     * from the session cache (a replayed event stream may be shorter than
     * the book). A miss schedules the chapter and answers null.
     */
    private fun completeGeometry(spineIdx: UInt): ChapterGeometry? {
        (readiness[spineIdx] as? ChapterReadiness.Complete)?.geometry
            ?.takeIf { it.pageCount > 0u }
            ?.let { return it }
        val geometry = runCatching { session.chapter(spineIdx) }.getOrNull() ?: return null
        if (latestGeneration != null && geometry.generation != latestGeneration) return null
        if (geometry.pageCount == 0u) return null
        readiness[spineIdx] = ChapterReadiness.Complete(geometry)
        return geometry
    }

    private fun setScene() {
        val count = displayedPageCount()
        rebaseInnerOffset(count)
        val edge = when {
            outerDisplacement == 0f -> null
            else -> {
                val toRight = outerDisplacement < 0f
                neighborEntry(toRight)
            }
        }
        canvas.setScene(
            PageScene(spineIdx, count, isRightToLeft, innerOffset, outerDisplacement, edge),
        )
    }

    private fun rebaseInnerOffset(pageCount: UInt) {
        if (isRightToLeft && scenePageCount > 0u && pageCount > scenePageCount) {
            innerOffset += (pageCount - scenePageCount).toFloat() * pageWidth
        }
        scenePageCount = pageCount
    }

    /**
     * The strip's page count. During a paging interaction it is frozen so
     * layout events cannot remap slots under the finger — except that in
     * LTR progression newly published pages append past the strip's end
     * without moving any existing offset, so growth is adopted live and a
     * drag can reach pages published during it. In RTL, growth would
     * rebase every offset; it stays deferred to [endPagingInteraction].
     */
    private fun displayedPageCount(): UInt {
        val frozen = interactionPageCount ?: return pageCountFor(spineIdx)
        if (isRightToLeft) return frozen
        val live = pageCountFor(spineIdx)
        if (live <= frozen) return frozen
        interactionPageCount = live
        return live
    }

    private fun neighborSpine(toRight: Boolean): UInt? {
        val delta = if (toRight) {
            if (isRightToLeft) -1 else 1
        } else {
            if (isRightToLeft) 1 else -1
        }
        val candidate = spineIdx.toLong() + delta
        return candidate.takeIf { it >= 0 && it < spineCount.toLong() }?.toUInt()
    }

    private fun isForward(toRight: Boolean): Boolean = toRight != isRightToLeft

    private fun neighborEntry(toRight: Boolean): PageNeighbor? {
        val neighbor = neighborSpine(toRight) ?: return null
        if (!neighborIsReady(toRight)) return null
        val page = when {
            neighbor in failedSpines -> 0u
            isForward(toRight) -> 0u
            else -> {
                val geometry = completeGeometry(neighbor) ?: return null
                geometry.pageCount - 1u
            }
        }
        return PageNeighbor(neighbor, page, toRight)
    }
}

File: apps/ios/Inkuna/Reader/Engine/EnginePagerSurface.swift
import UIKit

enum ChapterReadiness {
    case empty
    case partial(publishedPages: UInt32)
    case complete(geometry: ChapterGeometry)
}

/// The sole logical-page ↔ physical-slot conversion. The engine always
/// numbers pages in reading order; only the shell mirrors their strip slot
/// for RTL progression.
enum PageSlot {
    static func slot(for pageIdx: UInt32, pageCount: UInt32, rtl: Bool) -> UInt32 {
        guard pageIdx < pageCount else { return 0 }
        return rtl ? pageCount - pageIdx - 1 : pageIdx
    }

    static func pageIdx(for slot: UInt32, pageCount: UInt32, rtl: Bool) -> UInt32 {
        guard slot < pageCount else { return 0 }
        return rtl ? pageCount - slot - 1 : slot
    }

    static func offset(
        forPageIdx pageIdx: UInt32,
        pageCount: UInt32,
        rtl: Bool,
        pageWidth: CGFloat
    ) -> CGFloat {
        CGFloat(slot(for: pageIdx, pageCount: pageCount, rtl: rtl)) * pageWidth
    }
}

/// Engine-backed pager state. All FFI reads here are synchronous cache reads.
@MainActor
final class EnginePagerSurface: ReaderPagerSurface {
    private let session: ReaderSession
    private let canvas: EnginePageCanvas
    private var readiness: [UInt32: ChapterReadiness] = [:]
    private var failedSpines: Set<UInt32> = []
    private var neighborReadiness: [NeighborKey: Bool] = [:]
    private var latestGeneration: UInt64?
    private var pendingGeneration = false
    private var innerOffset: CGFloat = 0
    private var outerDisplacement: CGFloat = 0
    private var scenePageCount: UInt32 = 0
    private var interactionPageCount: UInt32?
    private var lastSettledPage: UInt32?

    private(set) var spineIdx: UInt32 = 0
    private(set) var pageIdx: UInt32 = 0
    var spineCount: UInt32 = 0
    var onPageSettled: ((UInt32, UInt32) -> Void)?
    var selectionActive = false

    init(session: ReaderSession, canvas: EnginePageCanvas) {
        self.session = session
        self.canvas = canvas
    }

    var isEngageable: Bool {
        guard !isBusy, canvas.isLaidOut else { return false }
        if failedSpines.contains(spineIdx) { return true }
        guard let latestGeneration else { return false }
        guard let list = try? session.page(spineIdx: spineIdx, pageIdx: pageIdx) else { return false }
        return list.generation == latestGeneration
    }

    var isBusy: Bool { pendingGeneration }
    var hasActiveSelection: Bool { selectionActive }
    var isRightToLeft: Bool { session.isRtl() }

    func display(spineIdx: UInt32, pageIdx: UInt32) {
        self.spineIdx = spineIdx
        self.pageIdx = pageIdx
        interactionPageCount = nil
        primeReadiness(for: spineIdx)
        let count = pageCount(for: spineIdx)
        innerOffset = PageSlot.offset(
            forPageIdx: pageIdx,
            pageCount: count,
            rtl: isRightToLeft,
            pageWidth: pageWidth
        )
        outerDisplacement = 0
        scenePageCount = count
        lastSettledPage = pageIdx
        neighborReadiness.removeAll()
        canvas.showUnreadablePlaceholder(failedSpines.contains(spineIdx))
        setScene()
        onPageSettled?(spineIdx, pageIdx)
    }

    func firstPageBecameReady(generation: UInt64, spineIdx: UInt32) {
        guard accept(generation: generation) else { return }
        let published = max(1, session.publishedPageCount(spineIdx: spineIdx))
        if case let .complete(geometry)? = readiness[spineIdx], geometry.generation == generation {
            // Completion may arrive before this hop reaches the main actor.
        } else {
            readiness[spineIdx] = .partial(publishedPages: published)
        }
        failedSpines.remove(spineIdx)
        neighborReadiness.removeAll()
        if spineIdx == self.spineIdx {
            pendingGeneration = false
            canvas.showUnreadablePlaceholder(false)
            setScene()
        }
    }

    func chapterBecameReady(generation: UInt64, spineIdx: UInt32) {
        guard accept(generation: generation), let geometry = try? session.chapter(spineIdx: spineIdx),
              geometry.generation == generation else { return }
        readiness[spineIdx] = .complete(geometry: geometry)
        failedSpines.remove(spineIdx)
        neighborReadiness.removeAll()
        if spineIdx == self.spineIdx {
            pendingGeneration = false
            canvas.showUnreadablePlaceholder(false)
            setScene()
        }
    }

    func chapterFailed(generation: UInt64, spineIdx: UInt32) {
        guard accept(generation: generation) else { return }
        readiness[spineIdx] = .empty
        failedSpines.insert(spineIdx)
        neighborReadiness.removeAll()
        if spineIdx == self.spineIdx {
            pendingGeneration = false
            canvas.showUnreadablePlaceholder(true)
            setScene()
        }
    }

    /// `updateLayout` returns `Void` in the generated bindings. The caller
    /// therefore passes 0 until the next engine callback supplies the real
    /// generation; no shell counter is created.
    func layoutInvalidated(generation: UInt64) {
        latestGeneration = generation == 0 ? nil : generation
        pendingGeneration = true
        readiness.removeAll()
        failedSpines.removeAll()
        neighborReadiness.removeAll()
        scenePageCount = 0
        interactionPageCount = nil
        lastSettledPage = nil
        canvas.invalidateAll()
    }

    func beginPagingInteraction() {
        let count = pageCount(for: spineIdx)
        rebaseInnerOffset(for: count)
        interactionPageCount = count
    }

    func endPagingInteraction() {
        interactionPageCount = nil
        setScene()
    }

    func innerMetrics() -> ReaderPagerStrip? {
        guard pageWidth > 0 else { return nil }
        let count = displayedPageCount()
        rebaseInnerOffset(for: count)
        // The strip's extent comes from the displayed count, so an
        // interaction's frozen count also freezes the range it can reach.
        let end = CGFloat(max(1, count) - 1) * pageWidth
        switch readiness[spineIdx] ?? .empty {
        case let .complete(geometry):
            if let latestGeneration, geometry.generation != latestGeneration { return nil }
            return ReaderPagerStrip(offset: innerOffset, range: 0 ... end, pageWidth: pageWidth)
        case .partial:
            return ReaderPagerStrip(
                offset: innerOffset,
                range: 0 ... max(innerOffset, end),
                pageWidth: pageWidth
            )
        case .empty:
            if failedSpines.contains(spineIdx) {
                return ReaderPagerStrip(offset: innerOffset, range: 0 ... 0, pageWidth: pageWidth)
            }
            return nil
        }
    }

    func setInnerOffset(_ x: CGFloat) {
        guard let metrics = innerMetrics() else { return }
        innerOffset = min(max(x, metrics.range.lowerBound), metrics.range.upperBound)
        setScene()
        let slot = UInt32((innerOffset / metrics.pageWidth).rounded())
        let settled = PageSlot.pageIdx(
            for: slot,
            pageCount: displayedPageCount(),
            rtl: isRightToLeft
        )
        if abs(innerOffset - PageSlot.offset(
            forPageIdx: settled,
            pageCount: displayedPageCount(),
            rtl: isRightToLeft,
            pageWidth: metrics.pageWidth
        )) < 0.01,
           settled != lastSettledPage {
            pageIdx = settled
            lastSettledPage = settled
            onPageSettled?(spineIdx, settled)
        }
    }

    /// The live outer offset — the exact inverse of `setOuterOffset`, so a
    /// pager that re-reads it mid-flight sees the boundary displacement
    /// currently on screen instead of the committed home.
    func outerMetrics() -> ReaderPagerStrip? {
        guard pageWidth > 0 else { return nil }
        let leftExists = neighborSpine(toRight: false) != nil
        let rightExists = neighborSpine(toRight: true) != nil
        return ReaderPagerStrip(
            offset: pageWidth - outerDisplacement,
            range: (leftExists ? 0 : pageWidth) ... (rightExists ? 2 * pageWidth : pageWidth),
            pageWidth: pageWidth
        )
    }

    func setOuterOffset(_ x: CGFloat) {
        guard let metrics = outerMetrics() else { return }
        let offset = min(max(x, metrics.range.lowerBound), metrics.range.upperBound)
        outerDisplacement = pageWidth - offset
        setScene()
    }

    func neighborIsReady(toRight: Bool) -> Bool {
        guard let neighbor = neighborSpine(toRight: toRight) else { return false }
        let key = NeighborKey(spineIdx: neighbor, toRight: toRight)
        if let ready = neighborReadiness[key] { return ready }
        let ready: Bool
        if failedSpines.contains(neighbor) {
            // A failed chapter occupies one placeholder page; it must stay
            // crossable in both directions or every chapter beyond it
            // becomes unreachable by paging.
            ready = true
        } else if isForward(toRight: toRight) {
            ready = session.publishedPageCount(spineIdx: neighbor) > 0
            if !ready {
                _ = try? session.chapter(spineIdx: neighbor)
            }
        } else {
            ready = completeGeometry(neighbor) != nil
        }
        neighborReadiness[key] = ready
        return ready
    }

    func commitBoundaryCrossing(toRight: Bool) -> Bool {
        guard let target = neighborSpine(toRight: toRight) else { return false }
        guard neighborIsReady(toRight: toRight) else { return false }
        let targetPage: UInt32
        if failedSpines.contains(target) || isForward(toRight: toRight) {
            targetPage = 0
        } else {
            guard let geometry = completeGeometry(target) else { return false }
            targetPage = geometry.pageCount - 1
        }
        spineIdx = target
        pageIdx = targetPage
        let count = pageCount(for: target)
        innerOffset = PageSlot.offset(
            forPageIdx: targetPage,
            pageCount: count,
            rtl: isRightToLeft,
            pageWidth: pageWidth
        )
        scenePageCount = count
        interactionPageCount = count
        outerDisplacement = 0
        lastSettledPage = pageIdx
        neighborReadiness.removeAll()
        canvas.showUnreadablePlaceholder(failedSpines.contains(target))
        setScene()
        onPageSettled?(spineIdx, pageIdx)
        return true
    }

    func pagePoint(fromCanvasPoint point: CGPoint) -> CGPoint? {
        canvas.pagePoint(spineIdx: spineIdx, pageIdx: pageIdx, from: point)
    }

    private var pageWidth: CGFloat { canvas.bounds.width }

    private struct NeighborKey: Hashable {
        let spineIdx: UInt32
        let toRight: Bool
    }

    private func pageCount(for spineIdx: UInt32) -> UInt32 {
        switch readiness[spineIdx] ?? .empty {
        case let .complete(geometry): geometry.pageCount
        case let .partial(publishedPages): max(publishedPages, session.publishedPageCount(spineIdx: spineIdx))
        case .empty: failedSpines.contains(spineIdx) ? 1 : 0
        }
    }

    /// Learns a spine's published prefix when a display gets ahead of its
    /// main-actor layout callback (for example after a recreated surface).
    private func primeReadiness(for spineIdx: UInt32) {
        guard readiness[spineIdx] == nil, !failedSpines.contains(spineIdx) else { return }
        let published = session.publishedPageCount(spineIdx: spineIdx)
        if published > 0 { readiness[spineIdx] = .partial(publishedPages: published) }
    }

    /// Complete geometry with at least one page, healing the readiness map
    /// from the session cache (a recreated surface starts with an empty map
    /// even when the session has already laid the chapter out). A miss
    /// schedules the chapter and answers nil.
    private func completeGeometry(_ spineIdx: UInt32) -> ChapterGeometry? {
        if case let .complete(geometry)? = readiness[spineIdx], geometry.pageCount > 0 {
            return geometry
        }
        guard let geometry = try? session.chapter(spineIdx: spineIdx) else { return nil }
        if let latestGeneration, geometry.generation != latestGeneration { return nil }
        guard geometry.pageCount > 0 else { return nil }
        readiness[spineIdx] = .complete(geometry: geometry)
        return geometry
    }

    private func accept(generation: UInt64) -> Bool {
        if let latestGeneration, generation != latestGeneration { return false }
        if latestGeneration == nil {
            latestGeneration = generation
            canvas.invalidate(generation: generation)
        }
        return true
    }

    private func setScene() {
        let count = displayedPageCount()
        rebaseInnerOffset(for: count)
        let edge: (spineIdx: UInt32, pageIdx: UInt32, toRight: Bool)?
        if outerDisplacement == 0 {
            edge = nil
        } else {
            let toRight = outerDisplacement < 0
            edge = neighborEntry(toRight: toRight)
        }
        canvas.setScene(PageScene(
            spineIdx: spineIdx,
            pageCount: count,
            rtl: isRightToLeft,
            innerOffset: innerOffset,
            outerDisplacement: outerDisplacement,
            neighborEdge: edge
        ))
    }

    private func rebaseInnerOffset(for pageCount: UInt32) {
        defer { scenePageCount = pageCount }
        guard isRightToLeft, scenePageCount > 0, pageCount > scenePageCount else { return }
        innerOffset += CGFloat(pageCount - scenePageCount) * pageWidth
    }

    /// The strip's page count. During a paging interaction it is frozen so
    /// layout events cannot remap slots under the finger — except that in
    /// LTR progression newly published pages append past the strip's end
    /// without moving any existing offset, so growth is adopted live and a
    /// drag can reach pages published during it. In RTL, growth would
    /// rebase every offset; it stays deferred to `endPagingInteraction()`.
    private func displayedPageCount() -> UInt32 {
        guard let frozen = interactionPageCount else { return pageCount(for: spineIdx) }
        guard !isRightToLeft else { return frozen }
        let live = pageCount(for: spineIdx)
        guard live > frozen else { return frozen }
        interactionPageCount = live
        return live
    }

    private func neighborSpine(toRight: Bool) -> UInt32? {
        let delta = toRight ? (isRightToLeft ? -1 : 1) : (isRightToLeft ? 1 : -1)
        let candidate = Int64(spineIdx) + Int64(delta)
        guard candidate >= 0, candidate < Int64(spineCount) else { return nil }
        return UInt32(candidate)
    }

    private func isForward(toRight: Bool) -> Bool { toRight != isRightToLeft }

    private func neighborEntry(toRight: Bool) -> (spineIdx: UInt32, pageIdx: UInt32, toRight: Bool)? {
        guard let neighbor = neighborSpine(toRight: toRight), neighborIsReady(toRight: toRight) else { return nil }
        if failedSpines.contains(neighbor) { return (neighbor, 0, toRight) }
        if isForward(toRight: toRight) {
            return (neighbor, 0, toRight)
        }
        guard let geometry = completeGeometry(neighbor) else { return nil }
        return (neighbor, geometry.pageCount - 1, toRight)
    }

}
